ts/card-info/index.ts

```diff
@@ -31,3 +33,19 @@ export async function cardInfo(
         props: { stats },
     });
 }
+
+export async function updateCardInfo(
```

Comment:
Instead of adding another entry to the "global" anki namespace, I think you could probably export a method in the Svelte instance instead. For example:

```diff
diff --git a/qt/aqt/browser/card_info.py b/qt/aqt/browser/card_info.py
index 26f99426c..e8e0fffd8 100644
--- a/qt/aqt/browser/card_info.py
+++ b/qt/aqt/browser/card_info.py
@@ -64,11 +64,11 @@ class CardInfoDialog(QDialog):
         self.setLayout(layout)
 
         self.web.eval(
-            f"let cardInfo = anki.cardInfo(document.getElementById('main'), {card_id}, true);"
+            f"const cardInfo = anki.cardInfo(document.getElementById('main'), {card_id}, true);"
         )
 
     def update_card(self, card_id: CardId) -> None:
-        self.web.eval(f"anki.updateCardInfo(cardInfo, {card_id}, true);")
+        self.web.eval(f"cardInfo.then((c) => c.updateCard({card_id}, true));")
 
     def reject(self) -> None:
         if self._on_close:
diff --git a/ts/card-info/CardInfo.svelte b/ts/card-info/CardInfo.svelte
index e1f8e1e97..90d72483c 100644
--- a/ts/card-info/CardInfo.svelte
+++ b/ts/card-info/CardInfo.svelte
@@ -7,6 +7,7 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
     import { Stats, unwrapOptionalNumber } from "../lib/proto";
     import { Timestamp, timeSpan, DAY } from "../lib/time";
     import Revlog from "./Revlog.svelte";
+    import { getCardStats } from "./lib";
 
     export let stats: Stats.CardStatsResponse;
 
@@ -93,6 +94,13 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 
     let statsRows: StatsRow[];
     $: statsRows = rowsFromStats(stats);
+
+    export function updateCard(cardId: number, includeRevlog: boolean) {
+        stats = await getCardStats(cardId);
+        if (!includeRevlog) {
+            stats.revlog = [];
+        }
+    }
 </script>
 
 <div class="container">
diff --git a/ts/card-info/index.ts b/ts/card-info/index.ts
index db972562c..61cd7d9ac 100644
--- a/ts/card-info/index.ts
+++ b/ts/card-info/index.ts
@@ -7,8 +7,6 @@ import { checkNightMode } from "../lib/nightmode";
 
 import CardInfo from "./CardInfo.svelte";
 
-const _updatingQueue: Promise<void> = Promise.resolve();
-
 export async function cardInfo(
     target: HTMLDivElement,
     cardId: number,
@@ -33,19 +31,3 @@ export async function cardInfo(
         props: { stats },
     });
 }
-
-export async function updateCardInfo(
-    cardInfo: Promise<CardInfo>,
-    cardId: number,
-    includeRevlog: boolean
-): Promise<void> {
-    _updatingQueue.then(async () => {
-        cardInfo.then(async (cardInfo) => {
-            const stats = await getCardStats(cardId);
-            if (!includeRevlog) {
-                stats.revlog = [];
-            }
-            cardInfo.$set({ stats });
-        });
-    });
-}
```

- the stats.revlog = [] inside the Svelte component shouldn't cause flicker, as rendering does not happen until the next render loop
- was the goal of updatingQueue to serialize access? I'm not sure we need it, but if we do, I think you'd probably need to return a new promise (there's no implicit return for => { ... }).
- You have the most experience here @hgiesel - any thoughts?

Reply:
What about storing a promise inside the Svelte component, and having the component doing its own fetching? The initial load case could then share the logic, creating the component and then calling updateCard() on it. Untested, but maybe something like the following?

```typescript
    let lastRequest: Promise<void> = Promise.resolve();

    export async function updateCard(cardId: number, includeRevlog: boolean) {
        lastRequest = lastRequest
            .then(() => getCardStats(cardId))
            .then((data) => {
                if (!includeRevlog) {
                    data.revlog = [];
                }
                stats = data;
            });
    }
```

Reply:
I was thinking something similar. And given your comment above, `cardInfo()` should probably be callable without a card id as well. So we could just as well skip the call to `updateCall()` and leave it to the consumer to set a card.
For clarity, I imagine an API like this:

```ts
/* build the page with some placeholder string */
const cardInfo = anki.cardInfo(document.getElementById('main'));
/* set a card */
cardInfo.then((c) => c.updateCard(123456789, true));")
```

Reply:
Yep, that's what I was thinking. An alternative approach would be to have cardId and the bool passed in as props and have the component call its internal update function each time the props change. It would be more ergonomic if this component was going to be embedded in a larger component, but since we're currently just calling it programmatically, I'm not sure we'd gain much from doing that at the moment.